Assets/MRTK/Providers/WindowsMixedReality/XRSDK/WindowsMixedRealityDeviceManager.cs

```diff
@@ -126,9 +140,91 @@ public override bool CheckCapability(MixedRealityCapability capability)
             return false;
         }
 
+#if HP_CONTROLLER_ENABLED
+        private MotionControllerWatcher motionControllerWatcher;
+
+        /// <summary>
+        /// Dictionary to capture all active HP controllers detected
+        /// </summary>
+        private readonly Dictionary<uint, MotionControllerState> trackedMotionControllerStates = new Dictionary<uint, MotionControllerState>();
+
+        private readonly Dictionary<uint, GenericXRSDKController> activeMotionControllers = new Dictionary<uint, GenericXRSDKController>();
+#endif
+
         #endregion IMixedRealityCapabilityCheck Implementation
 
+
         #region Controller Utilities
+#if HP_CONTROLLER_ENABLED
+        private static readonly ProfilerMarker GetOrAddControllerPerfMarker = new ProfilerMarker("[MRTK] WindwosMixedRealityXRSDKDeviceManager.GetOrAddController");
+
+        protected override GenericXRSDKController GetOrAddController(InputDevice inputDevice)
+        {
+            using (GetOrAddControllerPerfMarker.Auto())
+            {
+                GenericXRSDKController detectedController = base.GetOrAddController(inputDevice);
+                SupportedControllerType currentControllerType = GetCurrentControllerType(inputDevice);
+
+                // Add the Motion Controller state if it's an HPMotionController
+                if (currentControllerType == SupportedControllerType.HPMotionController)
+                {
+                    lock (trackedMotionControllerStates)
+                    {
+                        uint controllerId = GetControllerId(inputDevice);
+                        if (trackedMotionControllerStates.ContainsKey(controllerId))
+                        {
+                            ((HPMotionController)detectedController).MotionControllerState = trackedMotionControllerStates[controllerId];
+                        }
+                    }
+                }
+
+                return detectedController;
+            }
+        }
+
+        private void AddTrackedMotionController(object sender, MotionController motionController)
+        {
+            lock (trackedMotionControllerStates)
+            {
+                uint controllerId = GetControllerId(motionController);
+                trackedMotionControllerStates[controllerId] = new MotionControllerState(motionController);
+
+                if (activeMotionControllers.ContainsKey(controllerId))
+                {
+                    ((HPMotionController)activeMotionControllers[controllerId]).MotionControllerState = trackedMotionControllerStates[controllerId];
+                }
```

Comment:
Is it guaranteed that this active motion controller will be an `HPMotionController`? I think since we found out the input plug-in raises this event for all motion controllers (not just HP), we may want to update this to something like

```suggestion
                if (activeMotionControllers.ContainsKey(controllerId) && activeMotionControllers[controllerId] is HPMotionController hpController)
                {
                    hpController.MotionControllerState = trackedMotionControllerStates[controllerId];
                }
```

Reply:
Adding this check seems safer, I'll change it thanks!